src/group_theory/perm/concrete_cycle.lean

```diff
@@ -185,3 +186,219 @@ begin
+        pow_mod_card_support_cycle_of_self_apply, pow_add, mul_apply] }
+end
+
+lemma same_cycle.to_list_is_rotated {f : perm α} {x y : α} (h : same_cycle f x y) :
+  to_list f x ~r to_list f y :=
+begin
+  by_cases hx : x ∈ f.support,
+  { obtain ⟨_ | k, hk, hy⟩ := h.nat_of_mem_support _ hx,
+    { simp only [coe_one, id.def, pow_zero] at hy,
+      simp [hy] },
+    use k.succ,
+    rw [←to_list_pow_apply_eq_rotate, hy] },
+  { rw [to_list_eq_nil_iff.mpr hx, is_rotated_nil_iff', eq_comm, to_list_eq_nil_iff],
+    rwa ←h.mem_support_iff }
+end
+
+lemma pow_apply_mem_to_list_iff_mem_support {n : ℕ} :
+  (p ^ n) x ∈ p.to_list x ↔ x ∈ p.support :=
+begin
+  rw [mem_to_list_iff, and_iff_right_iff_imp],
+  refine λ _, same_cycle.symm _,
+  rw same_cycle_pow_left_iff
+end
+
+lemma to_list_form_perm_nil (x : α) :
+  to_list (form_perm ([] : list α)) x = [] :=
+by simp
+
+lemma to_list_form_perm_singleton (x y : α) :
+  to_list (form_perm [x]) y = [] :=
+by simp
+
+lemma to_list_form_perm_nontrivial (l : list α) (hl : 2 ≤ l.length) (hn : nodup l) :
+  to_list (form_perm l) (l.nth_le 0 (zero_lt_two.trans_le hl)) = l :=
+begin
+  have hc : l.form_perm.is_cycle := list.is_cycle_form_perm hn hl,
+  have hs : l.form_perm.support = l.to_finset,
+  { refine support_form_perm_of_nodup _ hn _,
+    rintro _ rfl,
+    simpa [nat.succ_le_succ_iff] using hl },
+  rw [to_list, hc.cycle_of_eq (mem_support.mp _), hs, card_to_finset, erase_dup_eq_self.mpr hn],
+  { refine list.ext_le (by simp) (λ k hk hk', _),
+    simp [form_perm_pow_apply_nth_le _ hn, nat.mod_eq_of_lt hk'] },
+  { simpa [hs] using nth_le_mem _ _ _ }
+end
+
+lemma to_list_form_perm_is_rotated_self (l : list α) (hl : 2 ≤ l.length) (hn : nodup l)
+  (x : α) (hx : x ∈ l):
+  to_list (form_perm l) x ~r l :=
+begin
+  obtain ⟨k, hk, rfl⟩ := nth_le_of_mem hx,
+  have hr : l ~r l.rotate k := ⟨k, rfl⟩,
+  rw form_perm_eq_of_is_rotated hn hr,
+  rw ←nth_le_rotate' l k k,
+  simp only [nat.mod_eq_of_lt hk, nat.sub_add_cancel hk.le, nat.mod_self],
+  rw [to_list_form_perm_nontrivial],
+  { simp },
+  { simpa using hl },
+  { simpa using hn }
+end
+
+lemma form_perm_to_list (f : perm α) (x : α) :
+  form_perm (to_list f x) = f.cycle_of x :=
+begin
+  by_cases hx : f x = x,
+  { rw [(cycle_of_eq_one_iff f).mpr hx, to_list_eq_nil_iff.mpr (not_mem_support.mpr hx),
+        form_perm_nil] },
+  ext y,
+  by_cases hy : same_cycle f x y,
+  { obtain ⟨k, hk, rfl⟩ := hy.nat_of_mem_support _ (mem_support.mpr hx),
+    rw [cycle_of_apply_apply_pow_self, list.form_perm_apply_mem_eq_next (nodup_to_list f x),
+        next_to_list_eq_apply, pow_succ, mul_apply],
+    rw mem_to_list_iff,
+    exact ⟨⟨k, rfl⟩, mem_support.mpr hx⟩ },
+  { rw [cycle_of_apply_of_not_same_cycle hy, form_perm_apply_of_not_mem],
+    simp [mem_to_list_iff, hy] }
+end
+
+lemma is_cycle.exists_unique_cycle {f : perm α} (hf : is_cycle f) :
+  ∃! (s : {s : cycle α // s.nodup ∧ s.nontrivial}), (s : cycle α).form_perm s.prop.left = f :=
```

Comment:
Are you sure the subtypes are helpful here?
```suggestion
  ∃! s : cycle α, s.nodup ∧ s.nontrivial ∧ s.form_perm s.prop.left = f :=
```
(and do I understand `∃!` well enough to have actually written an equivalent statement?)

Reply:
If you prefer the subtype version, I think it would still be handy to provide both.

Reply:
I will provide the conjuction-based form. The subtype-based form is for a later `fintype.choose`: https://github.com/leanprover-community/mathlib/pull/8226/files#diff-5ceb762aa98bd37660ef77ad0fea885f37a750216df0d0c25f7511050102c956R495-R496